hail/src/main/scala/is/hail/expr/ir/Parser.scala

```diff
@@ -0,0 +1,1009 @@
+import scala.util.parsing.input.Positional
+
+abstract class Token extends Positional {
+  def value: Any
+}
+
+final case class IdentifierToken(value: String) extends Token { }
+final case class StringToken(value: String) extends Token { }
+final case class IntegerToken(value: Long) extends Token { }
+final case class FloatToken(value: Double) extends Token { }
+final case class PunctuationToken(value: String) extends Token { }
+
+case class TokenIterator(a: Array[Token]) extends BufferedIterator[Token] {
+  private var pos = 0
+
+  def next(): Token = {
+    val x = a(pos)
+    pos += 1
+    x
+  }
+
+  def hasNext(): Boolean = pos < a.length
+
+  def head(): Token = {
+    assert(hasNext())
+    a(pos)
+  }
+}
+
+object IRLexer extends JavaTokenParsers {
+  val token: Parser[Token] =
+    "[()\\[\\]{}<>,:+@=]".r ^^ { p => PunctuationToken(p) } |
+      identifier ^^ { id => IdentifierToken(id) } |
+      float64_literal ^^ { d => FloatToken(d) } |
+      int64_literal ^^ { l => IntegerToken(l) } |
+      string_literal ^^ { s => StringToken(s) }
+
+  val lexer: Parser[Array[Token]] = rep(positioned(token)) ^^ { l => l.toArray }
+
+  def quotedLiteral(delim: Char, what: String): Parser[String] =
+    new Parser[String] {
+      def apply(in: Input): ParseResult[String] = {
+        var r = in
+
+        val source = in.source
+        val offset = in.offset
+        val start = handleWhiteSpace(source, offset)
+        r = r.drop(start - offset)
+
+        if (r.atEnd || r.first != delim)
+          return Failure(s"consumed $what", r)
+        r = r.rest
+
+        val sb = new StringBuilder()
+
+        val escapeChars = "\\bfnrt'\"`".toSet
+        var continue = true
+        while (continue) {
+          if (r.atEnd)
+            return Failure(s"unterminated $what", r)
+          val c = r.first
+          r = r.rest
+          if (c == delim)
+            continue = false
+          else {
+            sb += c
+            if (c == '\\') {
+              if (r.atEnd)
+                return Failure(s"unterminated $what", r)
+              val d = r.first
+              if (!escapeChars.contains(d))
+                return Failure(s"invalid escape character in $what", r)
+              sb += d
+              r = r.rest
+            }
+          }
+        }
+        Success(unescapeString(sb.result()), r)
+      }
+    }
+
+  override def stringLiteral: Parser[String] =
+    quotedLiteral('"', "string literal") | quotedLiteral('\'', "string literal")
+
+  def backtickLiteral: Parser[String] = quotedLiteral('`', "backtick identifier")
+
+  def identifier = backtickLiteral | ident
+
+  def string_literal: Parser[String] = stringLiteral
+
+  def int64_literal: Parser[Long] = wholeNumber.map(_.toLong)
+
+  def float64_literal: Parser[Double] =
+      "-inf" ^^ { _ => Double.NegativeInfinity } |
+      """-?\d+(\.\d+)?[eE][+-]?\d+""".r ^^ { _.toDouble } |
+      """-?\d*\.\d+""".r ^^ { _.toDouble } // FIXME: Is this correct??
```

Comment:
You can use the rule in the other parser if you put int before float above.  That's probably what I'd do (did).

Reply:
I changed both rules to just use the built-in JavaTokenParsers `floatingPointNumber`.